From 35dbb0e4ebcc07acecfd060ffc6ca076cf397920 Mon Sep 17 00:00:00 2001
From: Joshua Rogers <joshua@joshua.hu>
Date: Wed, 18 Mar 2026 17:08:03 +0100
Subject: [PATCH 1/3] handshake-checks: fix username comparison during
 rehandshake

This is definitely a security issue
subverting the GNUTLS_ALLOW_ID_CHANGE protection,
but its real-life exploitability is under question.

Reported-by: Joshua Rogers of AISLE Research Team <joshua@joshua.hu>
Fixes: #1808
Signed-off-by: Joshua Rogers <joshua@joshua.hu>
---
 lib/handshake-checks.c | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

--- a/lib/handshake-checks.c
+++ b/lib/handshake-checks.c
@@ -75,14 +75,18 @@ int _gnutls_check_id_for_change(gnutls_s
 		if (username == NULL)
 			return gnutls_assert_val(GNUTLS_E_INTERNAL_ERROR);
 
 		if (session->internals.saved_username &&
 		    session->internals.saved_username_size != -1) {
-			if (session->internals.saved_username_size == username_length &&
-			    strncmp(session->internals.saved_username, username, username_length)) {
-				_gnutls_debug_log("Session's PSK username changed during rehandshake; aborting!\n");
-				return gnutls_assert_val(GNUTLS_E_SESSION_USER_ID_CHANGED);
+			if (session->internals.saved_username_size !=
+				    username_length ||
+			    memcmp(session->internals.saved_username, username,
+				   username_length)) {
+				_gnutls_debug_log(
+					"Session's PSK username changed during rehandshake; aborting!\n");
+				return gnutls_assert_val(
+					GNUTLS_E_SESSION_USER_ID_CHANGED);
 			}
 		} else if (session->internals.saved_username == NULL &&
 			   session->internals.saved_username_size == -1) {
 			if (username_length > MAX_USERNAME_SIZE)
 				return gnutls_assert_val(GNUTLS_E_INTERNAL_ERROR);
